controller/controller_server/addons_server.py

```diff
@@ -0,0 +1,160 @@
+
+        volume_id_info = utils.get_volume_id_info(request.volume_id)
+        volume_internal_id = volume_id_info.internal_id
+
+        other_volume_id_info = utils.get_volume_id_info(request.target_volume_id)
+        other_volume_internal_id = other_volume_id_info.internal_id
+
+        other_system_id = request.parameters.get(config.PARAMETERS_SYSTEM_ID)
+
+        connection_info = utils.get_array_connection_info_from_secrets(request.secrets)
+        with get_agent(connection_info, volume_id_info.array_type).get_mediator() as mediator:
+            replication = mediator.get_replication(volume_internal_id, other_volume_internal_id, other_system_id)
+            if replication:
+                logger.info("deleting replication {} with system {}".format(replication.name,
+                                                                            other_system_id))
+                mediator.delete_replication(replication.name)
+            else:
+                logger.info("idempotent case. replication is already deleted with system {}".format(other_system_id))
+
+        return pb2.DisableVolumeReplicationResponse()
+
+    @staticmethod
+    def _ensure_volume_role_for_replication(mediator, replication, is_to_promote):
+        if is_to_promote:
+            if replication.is_primary:
+                logger.info("idempotent case. volume is already primary")
+            else:
+                logger.info("promoting volume for replication {}".format(replication.name))
+                mediator.promote_replication_volume(replication.name)
+        else:
+            if replication.is_primary:
+                logger.info("demoting volume for replication {}".format(replication.name))
+                mediator.demote_replication_volume(replication.name)
+            else:
+                logger.info("idempotent case. volume is already secondary")
+
+    def _ensure_volume_role(self, request, context, is_to_promote, response_type):
+        set_current_thread_name(request.volume_id)
+        method_name = "PromoteVolume" if is_to_promote else "DemoteVolume"
+        logger.info(method_name)
+        utils.validate_addons_request(request)
+
+        volume_id_info = utils.get_volume_id_info(request.volume_id)
+        volume_internal_id = volume_id_info.internal_id
+
+        other_volume_id_info = utils.get_volume_id_info(request.target_volume_id)
+        other_volume_internal_id = other_volume_id_info.internal_id
+
+        other_system_id = request.parameters.get(config.PARAMETERS_SYSTEM_ID)
+
+        connection_info = utils.get_array_connection_info_from_secrets(request.secrets)
+        with get_agent(connection_info, volume_id_info.array_type).get_mediator() as mediator:
+            replication = mediator.get_replication(volume_internal_id, other_volume_internal_id, other_system_id)
+            if not replication:
+                message = "could not find replication for volume internal id: {} " \
```

Comment:
`volume wwn/wwid/uuid..` are not the "internal" id, so we can't use those.
and IMO `volume id` is ambiguous, since it can be both `wwn/wwid/uuid..` and "an internal" vdisk id

Reply:
`volume storage id`?

Reply:
both `id`s are created by the storage, so that doesn't help differentiate between them either

Reply:
but the storage never calls wwn an id..

Reply:
![image](https://user-images.githubusercontent.com/57443811/132707903-6478eaa2-144f-44de-ae89-0108ea70e768.png)

Reply:
https://github.com/IBM/ibm-block-csi-driver/blob/d49fdfebe4bdf64c368e896a81a3b3bf5ef7bf57/controller/array_action/array_mediator_svc.py#L194

Reply:
ok, I'm forfeit.